### PR TITLE
Prevent alerts to be captured

### PR DESCRIPTION
**Pull request by maxbritto**
- Calls to the snapshot() function will now wait until all alert have
  disappeared before taking the screenshot (can be disabled with a
  boolean parameter on the snapshot() method)
- Adds a function to allow automatic dismissal of alerts that prevent
  actions or snapshots
- Copy the new SnapshotAlertManagement.swift file to the same place as
  the SnapshotHelper.swift file during original setup

### CODE REVIEW

**Comment by koenpunt** [comment]
Is the `waitForAlerts` argument necessary? Shouldn't it be that if `automaticallyDismissAlerts` is called, all alert are dismissed? And if someone doesn't want an alert to be dismissed, they simply call `stopDismissingAlerts`

**Comment by koenpunt** [comment]
Or if it stays, I think it should be named differently, because now it implies to wait for an alert to be shown, instead of the dismissal thereof

**Comment by maxbritto** [comment]
I'm not sure because `waitForAlerts` actually does wait but doesn't dismiss the alert by itself. It does the same thing as `waitForLoadingIndicator`. It lets the developer chose its own way of dismissing the alert : he can take care of it on it's own if the alert matters or call `automaticallyDismissAlerts` to automatically dismiss any alert that gets in the way.

**Comment by maxbritto** [comment]
Regarding your second comment, I see what you mean but then `waitForLoadingIndicator` should also be renamed.

**Comment by koenpunt** [comment]
> Regarding your second comment, I see what you mean but then waitForLoadingIndicator should also be renamed.

You're right.

> I'm not sure because waitForAlerts actually does wait but doesn't dismiss the alert by itself. It does the same thing as waitForLoadingIndicator. 

`waitForLoading` indicator waits for the loading indicator to disappear, but you say that `waitForAlerts` doesn't dismiss alert? That might be confusing to users

**Comment by maxbritto** [comment]
I tried to follow the pattern of waitForLoading and do the same : I just wait for the alert to disappear. This is based on the assumption that many developer are probably doing meaningful things with their alerts and do not want us to dismiss them as they appear.
But I've also added a print if we have been waiting for more than 2 seconds and an alert is still here : it explains how to enable the automatic dismissal if they want it. I also added documentation in setup.rb

File: lib/assets/SnapshotAlertManager.swift [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        var waitCounter = 0
+        while (query.count > 0) {
+            sleep(1)
+            waitCounter++
+            print("Found an alert... waiting for it to disappear")
+            if waitCounter > 2 {
+              if alertInterruptionMonitor == nil {
+                  print("The SnapshotAlertManager can automatically hide alerts that will prevent actions or screenshots. Just call this method to enable it : automaticallyDismissAlerts(testCase:XCTestCase)")
+              } else {
+                  XCUIApplication().tap() //UIInterruptionMonitor will only trigger if an action is prevented by the alert so we generate one if the user script hasn't after 2 seconds
```
Comment (makadaw) [inline lib/assets/SnapshotAlertManager.swift]: XCUIApplication().tap() may tap on some control if app do not display alert

Comment (maxbritto) [inline lib/assets/SnapshotAlertManager.swift]: We arrive here only after an alert has been present for at least 2 seconds so the app will be displaying the alert

**Comment by i2amsam** [comment]
Hey @maxbritto thank you for this!  

I'm not going to merge this for now, though I think this an interesting problem, I'm concerned about adding additional dependencies and the migration path for existing setups. This could be a good candidate for getting "local helpers" or "local action" for some sort of plugin system in `snapshot` so the helpers can be easily extended.

**Comment by maxbritto** [comment]
I've never seen any plugin system for fastlane/snapshot. If there is one I'd be willing to refactor this as a plug-in. But if you're asking me to help building the plugin system I'm not sure how efficient I would be since I've never done any Ruby.

**Pull request closed without merging**